packages/ui/src/server/components/vcs/index.ts

```diff
@@ -0,0 +1,59 @@
+import {VCSSettings} from '../../../shared/ui-settings';
+import {GithubApi} from './GithubApi';
+import {GitlabApi} from './GitlabApi';
+import {AxiosError, AxiosResponse} from 'axios';
+import {Request} from '@gravity-ui/expresskit';
+import {AppConfig} from '@gravity-ui/nodekit';
+
+export interface VcsApi {
+    token: string;
+    api: string;
+    vcsId: string;
+
+    getRequestHeaders(): {Authorization: string};
+    getRepositories(): Promise<unknown | AxiosError>;
+    getBranches(projectId: string): Promise<string[] | AxiosError>;
+    getRepositoryContent(
+        projectId: string,
+        branch: string,
+        path: string,
+    ): Promise<unknown[] | AxiosError>;
+    getFileContent(
+        projectId: string,
+        branch: string,
+        path: string,
+    ): Promise<AxiosResponse<NodeJS.ReadableStream> | AxiosError>;
+}
+
+export class VcsFactory {
+    static createVcs(vcs: VCSSettings, token: string) {
+        const apiConfig = {token, api: vcs.api, vcsId: vcs.id};
+        switch (vcs.type) {
+            case 'github':
+                return new GithubApi(apiConfig);
+            case 'gitlab':
+                return new GitlabApi(apiConfig);
+            default:
+                throw new Error(`Unsupported VCS type: ${vcs.type}`);
+        }
+    }
+
+    static getVcsSettings(req: Request): VCSSettings[] {
+        const {uiSettings} = req.ctx.config as AppConfig;
+        if (!('vcsSettings' in uiSettings)) {
+            throw new Error('vcsSettings is not found in UISettings');
```

Comment:
It is better to use app-specific error type to send correct `http status code 400`.
```ts
// in server/utils/index.ts
export class BadRequest extends Error {}

export function sendError(res: Response, error: unknown, errCode = 500) {
    const errData = prepareErrorToSend(error);
    const statusCode = error instanceof BadRequest ? 400 : errCode;
    return res.set('content-type', 'application/json').status(statusCode).send(errData);
}

// here
throw new BadRequest(...)
```